### ISSUE TITLE
cudnn RNN memory consumption is not coherent

### ISSUE THREAD

**Issue by stefbraun** [pytorch/pytorch#2895]
Hi, I am experimenting with cudnn RNNs and i found that in terms of memory consumption, the order across cell types is as following (low memory usage to high memory usage):
`RNN` < **`LSTM`** < `GRU` (measured).

As GRU has less parameters than LSTM, it would be expected that the order is
`RNN` < `GRU` < **`LSTM`** (expected).

The script attached at the end enables reproduction (see output below). Two questions arise:
1. Why does GRU consume more memory although it has less parameters than LSTM?
2. Why is more memory consumed in the 2nd and 3rd epoch?
```
Epoch 0 - RNN - 761930 params ::: GeForce GTX 980 Ti: 5153.75 MB free, 924.5625 MB used, 6078.3125 MB total
Epoch 1 - RNN - 761930 params ::: GeForce GTX 980 Ti: 5150.75 MB free, 927.5625 MB used, 6078.3125 MB total
Epoch 2 - RNN - 761930 params ::: GeForce GTX 980 Ti: 5150.75 MB free, 927.5625 MB used, 6078.3125 MB total
```
```
Epoch 0 - GRU - 2279370 params ::: GeForce GTX 980 Ti: 3910.125 MB free, 2168.1875 MB used, 6078.3125 MB total
Epoch 1 - GRU - 2279370 params ::: GeForce GTX 980 Ti: 3086.125 MB free, 2992.1875 MB used, 6078.3125 MB total
Epoch 2 - GRU - 2279370 params ::: GeForce GTX 980 Ti: 3086.125 MB free, 2992.1875 MB used, 6078.3125 MB total
```
```
Epoch 0 - LSTM - 3038090 params ::: GeForce GTX 980 Ti: 4133.5 MB free, 1944.8125 MB used, 6078.3125 MB total
Epoch 1 - LSTM - 3038090 params ::: GeForce GTX 980 Ti: 3431.5 MB free, 2646.8125 MB used, 6078.3125 MB total
Epoch 2 - LSTM - 3038090 params ::: GeForce GTX 980 Ti: 3431.5 MB free, 2646.8125 MB used, 6078.3125 MB total
```
```
from timeit import default_timer as timer
import torch.cuda as cutorch
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.autograd import Variable
from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence, PackedSequence
from pynvml import *

# nvml
nvmlInit()
for i in range(nvmlDeviceGetCount()):
    handle = nvmlDeviceGetHandleByIndex(i)

# Set seeds
np.random.seed(11)
torch.manual_seed(11)
torch.cuda.manual_seed(11)

# Parameters
batch_size = 25
max_length = 1000
features = 123
num_classes = 10

arch = 'LSTM'
num_layers = 4
hidden_size = 320

epochs = 3
# Toy data
batch = torch.rand(max_length, batch_size, features)
labels = torch.from_numpy((np.random.randint(low=0, high=num_classes - 1, size=batch_size, dtype=np.int32)))

# Network
class rnn(nn.Module):
    def __init__(self, features, num_layers, hidden_size, arch, num_classes):
        super(rnn, self).__init__()
        if arch == 'RNN':
            self.rnn = nn.RNN(input_size=features, num_layers=num_layers, hidden_size=hidden_size, bias=True)
        if arch == 'GRU':
            self.rnn = nn.GRU(input_size=features, num_layers=num_layers, hidden_size=hidden_size, bias=True)
        if arch == 'LSTM':
            self.rnn = nn.LSTM(input_size=features, num_layers=num_layers, hidden_size=hidden_size, bias=True)
        self.fc = nn.Linear(hidden_size, num_classes)
    def forward(self, x):
        x, _ = self.rnn(x)
        x = self.fc(x[-1,:,:])# RNN
        return x

# Net, optimizer
net = rnn(features, num_layers, hidden_size, arch=arch, num_classes = num_classes).cuda()
optimizer = optim.Adam(net.parameters())
criterion = nn.CrossEntropyLoss()

# Parameter count
params = 0
for param in list(net.parameters()):
    sizes = 1
    for el in param.size():
        sizes = sizes * el
    params += sizes

for epoch in range(epochs):
    batch_var = Variable(batch).cuda()
    labels_var = Variable(labels).cuda()
    optimizer.zero_grad()
    output = net(batch_var)
    loss = criterion(output, labels_var.long())
    loss.backward()
    optimizer.step()

    meminfo = nvmlDeviceGetMemoryInfo(handle)
    print('Epoch {} - {} - {} params ::: {}: {} MB free, {} MB used, {} MB total'.format(epoch, arch,params,
        nvmlDeviceGetName(handle),
        meminfo.free / 1024. ** 2, meminfo.used / 1024. ** 2, meminfo.total / 1024. ** 2))
```

**Comment by apaszke** [comment]
It all boils down to what cuDNN does. Maybe it uses a larger workspace for GRUs to make them faster? @ngimel?

The memory increase you see is because of Python scoping rules. If you put the body of the epoch loop in a function or added `del loss` at the end, you'd see that the memory usage is constant.

**Comment by stefbraun** [comment]
@soumith why has this been closed? The cudnn behaviour is not comprehensible and leads to out-of-memory situations.

**Comment by soumith** [comment]
I'll wait for @ngimel to respond, but i closed it because this seems like an inactionable issue. That's what I determined from @apaszke 's reply.

**Comment by apaszke** [comment]
Yup. I just wanted to see if @ngimel can check if it's consistent with what's inside cuDNN. If it is, we won't be taking any action on this issue

**Comment by ngimel** [comment]
Reserve size returned by cudnn is indeed larger for GRU in this example (864000000 vs 736000000). I could not reproduce overall higher mem usage (see my output below), but in my case only 10 MB showed up as used, so must be something wrong with my pynvml. However, even properly working nvml is not the right tool to check memory usage, because memory cached in mem allocator that can be freed when needed will show up as used in nvml. 
```
workspace size 13169920
reserve size 864000000
Epoch 0 - GRU - 2279370 params ::: b'Tesla P100-SXM2-16GB': 16266.25 MB free, 10.0 MB used, 16276.25 MB total
workspace size 13169920
reserve size 864000000
Epoch 1 - GRU - 2279370 params ::: b'Tesla P100-SXM2-16GB': 16266.25 MB free, 10.0 MB used, 16276.25 MB total
workspace size 13169920
reserve size 864000000
Epoch 2 - GRU - 2279370 params ::: b'Tesla P100-SXM2-16GB': 16266.25 MB free, 10.0 MB used, 16276.25 MB total
root@79c48f5ef1c6:/workspace/ALL/playground# vim rnn
rnnleak.py  rnnmem.py   
root@79c48f5ef1c6:/workspace/ALL/playground# vim rnnmem.py 
root@79c48f5ef1c6:/workspace/ALL/playground# python rnnmem.py 
workspace size 17730560
reserve size 736000000
Epoch 0 - LSTM - 3038090 params ::: b'Tesla P100-SXM2-16GB': 16266.25 MB free, 10.0 MB used, 16276.25 MB total
workspace size 17730560
reserve size 736000000
Epoch 1 - LSTM - 3038090 params ::: b'Tesla P100-SXM2-16GB': 16266.25 MB free, 10.0 MB used, 16276.25 MB total
workspace size 17730560
reserve size 736000000
Epoch 2 - LSTM - 3038090 params ::: b'Tesla P100-SXM2-16GB': 16266.25 MB free, 10.0 MB used, 16276.25 MB total
```

**Comment by ngimel** [comment]
This is indeed expected, because of the different structure of the GRU cell compared to LSTM (in LSTM, one just adds the results of input and recurrent matrix multiplies, in GRU there's a newgate), so for efficient computation of weight gradient, GRU has to store 6 values per input value for LSTM just 4 is enough.  Total reserve space for GRU is 7 values per input value, and for LSTM it is 6.

**Comment by soumith** [comment]
Stef, let us know if you have any follow up questions / would like us to take this issue in a particular direction.

**Comment by stefbraun** [comment]
Thank you for the investigation and explanation, from my side there are no further questions.
